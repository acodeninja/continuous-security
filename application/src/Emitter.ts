import EventEmitter2 from 'eventemitter2';

export class Emitter implements EventEmitter {
  private eventEmitter: EventEmitter2;

  constructor() {
    this.eventEmitter = new EventEmitter2({wildcard: true});
  }

  emit(event: string, ...values: Array<unknown>): void {
    this.eventEmitter.emit(event, ...values);
  }

  on(event: string, listener: (...values: Array<unknown>) => void): void {
    this.eventEmitter.on(event, listener);
  }
}
