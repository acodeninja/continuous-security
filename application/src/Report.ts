import {template, TemplateExecutor} from 'lodash';
import {basename, resolve} from 'path';
import {readFile, writeFile} from 'fs/promises';
import {Converter as Showdown} from 'showdown';
import {promisify} from 'util';
import {exec} from 'child_process';

import {Emitter} from './Emitter';
import {translate} from './DataSources/Translations';
import {References} from './DataSources/References';
import {makeTemporaryFolder} from './Helpers/Files';

import MarkdownTemplate from './assets/report.markdown.template.md';
import PDFTemplate from './assets/report.pdf.template.md';
import HTMLTemplate from './assets/report.html.template.md';
import HTMLTemplateWrapper from './assets/report.html.wrapper.html';

export class Report {
  private readonly templates: Record<string, TemplateExecutor>;
  private reports: Array<ScanReport> = [];
  private cached: ReportOutput;
  private emitter: Emitter;
  private references: References;
  private severityOrder = ['critical', 'high', 'moderate', 'low', 'info', 'unknown'];
  private reportFunctions: Record<string, (...args: Array<unknown>) => unknown> = {
    groupBy: (list: Array<unknown>, grouping: string) => {
      const groupedIssues =
        [{}, ...list].reduce((group: Record<string, Array<unknown>> = {}, item) => {
          if (item) {
            group[item[grouping]] = group[item[grouping]] ?? [];
            group[item[grouping]].push(item);
          }
          return group;
        });

      const asEntries = Object.entries(groupedIssues);

      asEntries.sort((a, b) => this.severityOrder.indexOf(a[0]) - this.severityOrder.indexOf(b[0]));

      return Object.fromEntries(asEntries);
    },
    capitalise: (words: string) =>
      words.split(' ')
        .map(w => w.charAt(0).toUpperCase() + w.slice(1))
        .join(' '),
  };

  constructor(emitter: Emitter) {
    this.emitter = emitter;
    this.references = new References(emitter);
    this.templates = {
      markdown: template(MarkdownTemplate),
      html: template(HTMLTemplate),
      pdf: template(PDFTemplate),
    };
  }

  addScanReport(report: ScanReport): void {
    this.reports.push(report);
  }

  async toObject(): Promise<ReportOutput> {
    if (this.cached) return this.cached;

    const issues = await this.getIssues();

    const overviewOfIssues = issues.map(i => i.references).flat()
      .filter(r => !!r.dataSourceSpecific.cwe)
      .filter((r, i, all) => all.map(a => a.label).indexOf(r.label) === i);

    const summaryImpacts: Record<string, Array<string>> = {};

    overviewOfIssues
      .map(({dataSourceSpecific}) =>
        dataSourceSpecific?.cwe?.consequences?.map(({scopeImpacts}) =>
          scopeImpacts,
        ),
      )
      .flat(2)
      .forEach(si => {
        summaryImpacts[si.scope] = summaryImpacts[si.scope] ?? [];
        if (si.impact && !summaryImpacts[si.scope].includes(si.impact)) {
          summaryImpacts[si.scope].push(si.impact);
        }
      });

    overviewOfIssues.sort((a, b) =>
      parseInt(a.label.replace('CWE-', '')) - parseInt(b.label.replace('CWE-', '')));

    const severities = issues.map(i => i.severity);

    this.cached = JSON.parse(JSON.stringify({
      title: `Security Report for ${basename(process.cwd())}`,
      date: new Date(),
      summaryImpacts: Object.entries(summaryImpacts).map(([scope, impacts]) => ({scope, impacts})),
      overviewOfIssues,
      issues,
      counts: {
        critical: severities.filter(s => s === 'critical').length,
        high: severities.filter(s => s === 'high').length,
        moderate: severities.filter(s => s === 'moderate').length,
        low: severities.filter(s => s === 'low').length,
        info: severities.filter(s => s === 'info').length,
        unknown: severities.filter(s => s === 'unknown').length,
        total: severities.length,
      },
    }), (key, value) => {
      if (
        String(value).length === 24 &&
        String(value).match(/\d{4}-\d{2}-\d{2}T(\d{2}:){2}\d{2}.\d{3}Z/)
      ) return new Date(value);

      if (key === 'code' || key === 'target' || key === 'name') return value;

      return typeof value === 'string' ? translate(value) : value;
    });

    return this.cached;
  }

  async toJSON(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    return ['json', Buffer.from(JSON.stringify(report, null, 2))];
  }

  async toMarkdown(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    return [
      'md',
      Buffer.from(this.templates.markdown({...report, functions: this.reportFunctions})),
    ];
  }

  async toHTML(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    const convert = new Showdown();
    const converted = convert.makeHtml(this.templates.html({
      ...report,
      functions: this.reportFunctions,
    }));

    const html = HTMLTemplateWrapper.toString().replace('%%REPORT%%', converted);

    return ['html', Buffer.from(html)];
  }

  async toPDF(): Promise<[string, Buffer]> {
    const htmlCacheLocation = await makeTemporaryFolder('html');

    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    const convert = new Showdown();
    const converted = convert.makeHtml(this.templates.pdf({
      ...report,
      functions: this.reportFunctions,
    }));

    const html = HTMLTemplateWrapper.toString().replace('%%REPORT%%', converted);

    await writeFile(resolve(htmlCacheLocation, 'report.html'), html);

    const pdfPath = resolve(htmlCacheLocation, 'report.pdf');
    const htmlPath = resolve(htmlCacheLocation, 'report.html');

    // chrome --headless --disable-gpu --print-to-pdf https://www.chromestatus.com/

    try {
      await promisify(exec)(
        `google-chrome \
        --headless \
        --disable-gpu \
        --no-margins \
        --no-pdf-header-footer \
        --print-to-pdf="${pdfPath}" \
        ${htmlPath}`,
      );
    } catch (e) {
      console.log(e);
    }

    return ['pdf', Buffer.from(await readFile(pdfPath))];
  }

  async getReport(type: 'markdown' | 'json' | 'html' | 'pdf'): Promise<[string, Buffer]> {
    this.emitter.emit('report:started', `generating output report in ${type}`);
    switch (type) {
    case 'markdown':
      return await this.toMarkdown();
    case 'html':
      return await this.toHTML();
    case 'pdf':
      return await this.toPDF();
    case 'json':
      return await this.toJSON();
    }
  }

  private async getIssues(): Promise<Array<ReportOutputIssue>> {
    const withFoundBy = this.reports.map(r => r.issues.map(issue => ({
      ...issue,
      foundBy: r.scanner,
    }))).flat();

    const withExpandedReferences = [];

    for (const issue of withFoundBy) {
      withExpandedReferences.push({
        ...issue,
        references: issue.references ? await this.references.getAll(issue.references) : [],
      });
    }

    const withHighestSeverities = withExpandedReferences.map(i => {
      const severities = i.references
        .map((r: ReportOutputIssueReference) => r.dataSourceSpecific?.osv?.severity)
        .concat(i.severity)
        .filter((s: ReportOutputIssue['severity'] | undefined) => !!s);

      severities.sort((a, b) => this.severityOrder.indexOf(a) - this.severityOrder.indexOf(b));

      return {...i, severity: severities[0]};
    });

    return await Promise.all(withHighestSeverities.map(async issue => {
      return {
        ...issue,
        extracts: issue.extracts ? await Promise.all(issue.extracts.map(async extract => {
          const code = (await readFile(extract.path)).toString().split('\n');
          const start = Math.max(parseInt(extract.lines[0]) - 3, 0);
          const end = Math.min(parseInt(extract.lines[0]) + 2, code.length);
          const lines = new Array(code.slice(start, end).length)
            .fill(null)
            .map((_, i) => start + i + 1);

          return {
            ...extract,
            lines: extract.lines.map(line => `${line}`),
            code: code.slice(start, end).map((l, i) => {
              const length = lines[lines.length - 1].toString().length;
              return `${lines[i].toString().padStart(length)}| ${l}`;
            }).join('\n'),
          };
        })) : undefined,
      };
    }));
  }
}
