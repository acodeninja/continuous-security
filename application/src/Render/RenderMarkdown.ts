import {template} from 'lodash';

import {Report} from '../Report';
import {Emitter} from '../Emitter';
import {groupToObjectBy} from '../Helpers/Arrays';
import {capitalise} from '../Helpers/Strings';

import MarkdownTemplate from '../assets/report.markdown.template.md';
import {timezone, toDate, toTime} from '../Helpers/Dates';

export class RenderMarkdown {
  private report: Report;
  private emitter: Emitter;

  constructor(report: Report, emitter: Emitter) {
    this.report = report;
    this.emitter = emitter;
  }

  async render(): Promise<Buffer> {
    this.emitter.emit('report:render:markdown:started');

    const report = await this.report.toObject();
    const output = template(MarkdownTemplate)({...report, functions: {
      groupToObjectBy,
      capitalise,
      toDate,
      toTime,
      timezone,
    }});

    this.emitter.emit('report:render:markdown:finished');

    return Buffer.from(output);
  }
}
