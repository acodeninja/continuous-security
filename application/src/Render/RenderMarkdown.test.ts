import {Report} from '../Report';
import {Emitter} from '../Emitter';
import {RenderMarkdown} from './RenderMarkdown';

import {emptyReport, fullReport} from '../../tests/fixtures/Reports';

describe('RenderMarkdown', () => {
  describe.each([
    ['empty', emptyReport],
    ['full', fullReport],
  ])('an %s report', (name, reportTemplate) => {
    const onReportRenderStarted = jest.fn();
    const onReportRenderFinished = jest.fn();
    const emitter = new Emitter();
    const report = new Report(emitter);
    const render = new RenderMarkdown(report, emitter);
    const reportToObject = jest.spyOn(report, 'toObject');
    let output: Buffer;

    beforeAll(async () => {
      emitter.on('report:render:markdown:started', onReportRenderStarted);
      emitter.on('report:render:markdown:finished', onReportRenderFinished);
      reportToObject.mockResolvedValue(reportTemplate);
      output = await render.render();
    });

    test('emits event report:render:markdown:started', () => {
      expect(onReportRenderStarted).toHaveBeenCalled();
    });

    test('emits event report:render:markdown:finished', () => {
      expect(onReportRenderFinished).toHaveBeenCalled();
    });

    test('calls Report.toObject', () => {
      expect(reportToObject).toHaveBeenCalledWith();
    });

    test('returns the expected report', () => {
      expect(output.toString()).toMatchSnapshot();
    });
  });
});
