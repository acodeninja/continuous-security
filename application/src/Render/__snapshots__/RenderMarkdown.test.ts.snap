// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderMarkdown an empty report returns the expected report 1`] = `
"# Empty Report

# Summary

This security report was conducted on 01/04/2020 at 01:30:10 (UTC+0).
A total of 0 issue(s) were found, 0 of which may require immediate attention.

This report is produced by running automated security scanning tools, which will likely not detect
all vulnerabilities present. **It is not a replacement for a manual analysis of the application**.

The following technical impacts may arise if an adversary successfully exploits one of the issues found by this scan.


# Statistics

This report found issues with the following severities.

**Critical**: 0 | **High** 0 | **Medium** 0 | **Low** 0 | **Informational** 0 | **Unknown** 0

To gain a better understanding of the severity levels please see [the appendix](#what-are-severity-levels).

## Contents

* [Summary](#summary)
* [Issue Statistics](#statistics)
* [Overview of Issues](#overview-of-issues)
* [Vulnerabilities](#vulnerabilities)
* [Additional Information](#additional-information)
    * [What are severity levels?](#what-are-severity-levels)

# Overview of Issues


# Vulnerabilities



# Additional Information

## What are severity levels?

Issue severity is scored using the [Common Vulnerability Scoring System](https://www.first.org/cvss/) (CVSS) where
such data is available. Severity levels do not represent the risk associated with an issue as risk depends on your
specific context. Severity scoring does however give an indication of the ease of exploitation and potential scope of an
attacks effect on an application.

### Critical

Exploitation will likely lead to an attacker gaining administrative access to the application and infrastructure that
supports it. Exploiting critical vulnerabilities is usually trivial and will generally not require prior access to the
application. **A development team should aim to resolve these issues immediately by mitigating or directly resolving the
issue**.

### High

Exploitation could lead to an attacker gaining elevated access to the application and the infrastructure that supports
it. It is likely that an attacker will not find exploitation trivial. Such exploitation could lead to significant data
loss or downtime.

### Medium

Exploitation could lead to an attacker gaining limited access to the application. Exploiting vulnerabilities may require
an attacker to manipulate users to gain access to their credentials. Such exploitation could lead to limited data loss
or downtime.

### Low

Exploitation will likely have very little impact on the application, and it is unlikely that an attacker will gain any
meaningful access to the application. Exploiting an issue of this severity will potentially require physical access to
the infrastructure that supports the application.

### Informational

While not part of the CVSS scoring specification, several security analysis tools use this severity level to indicate
that an issue is a matter of best practice. It is extremely unlikely that issues with this severity will lead to an
attacker gaining access to any application components.

### Unknown

This severity level is used when the analysis tool used to perform a scan of the application does not associate any kind
of severity level with the issues or vulnerabilities it finds. Issues with an unknown severity should be investigated by
application developers and project stakeholders to establish the ease of exploitation, scope of any potential impact and
the specific risks associated.
"
`;

exports[`RenderMarkdown an full report returns the expected report 1`] = `
"# Security Report for application

# Summary

This security report was conducted on 01/04/2020 at 01:30:10 (UTC+0).
A total of 1 issue(s) were found, 0 of which may require immediate attention.

This report is produced by running automated security scanning tools, which will likely not detect
all vulnerabilities present. **It is not a replacement for a manual analysis of the application**.

The following technical impacts may arise if an adversary successfully exploits one of the issues found by this scan.

* **Availability**: DoS: Crash, Exit, or Restart
* **Confidentiality**: Read Application Data

# Statistics

This report found issues with the following severities.

**Critical**: 0 | **High** 0 | **Medium** 0 | **Low** 0 | **Informational** 0 | **Unknown** 1

To gain a better understanding of the severity levels please see [the appendix](#what-are-severity-levels).

## Contents

* [Summary](#summary)
* [Issue Statistics](#statistics)
* [Overview of Issues](#overview-of-issues)
    * [Uncaught Exception](#CWE-248)
* [Vulnerabilities](#vulnerabilities)
    * [Unknown (1)](#unknown-severity)
* [Additional Information](#additional-information)
    * [What are severity levels?](#what-are-severity-levels)

# Overview of Issues

<a id="CWE-248"></a>
## Uncaught Exception

An exception is thrown from a function, but it is not caught.


When an exception is not caught


### Consequences

Using a vulnerability of this type an attacker may be able to affect the system in the following ways.

* **Availability**: DoS: Crash, Exit, or Restart
* **Confidentiality**: Read Application Data

> An uncaught exception


For more information see [CWE-248](https://cwe.mitre.org/data/definitions/248.html).


# Vulnerabilities

## Unknown Severity

### test-issue-title 

**Severity**: [Unknown](#unknown) | **Type**: code smell | **Fix**: Unknown | **Found By**: [test-scanner](https://www.npmjs.com/package/test-scanner)

test-issue-description

#### Evidence

The following evidence of this vulnerability was found in the application.

\`application/src/Report\` (starting on line: 2)
\`\`\`python
import
\`\`\`

\`application/src/Report\` (starting on line: 2)
\`\`\`python
import
\`\`\`

\`application/src/Report\` (starting on line: 2)
\`\`\`python
import
\`\`\`

\`application/src/Report\` (starting on line: 2)
\`\`\`python
import
\`\`\`

\`application/src/Report\` (starting on line: 2)
\`\`\`python
import
\`\`\`


##### References

[CWE-248](#CWE-248)



# Additional Information

## What are severity levels?

Issue severity is scored using the [Common Vulnerability Scoring System](https://www.first.org/cvss/) (CVSS) where
such data is available. Severity levels do not represent the risk associated with an issue as risk depends on your
specific context. Severity scoring does however give an indication of the ease of exploitation and potential scope of an
attacks effect on an application.

### Critical

Exploitation will likely lead to an attacker gaining administrative access to the application and infrastructure that
supports it. Exploiting critical vulnerabilities is usually trivial and will generally not require prior access to the
application. **A development team should aim to resolve these issues immediately by mitigating or directly resolving the
issue**.

### High

Exploitation could lead to an attacker gaining elevated access to the application and the infrastructure that supports
it. It is likely that an attacker will not find exploitation trivial. Such exploitation could lead to significant data
loss or downtime.

### Medium

Exploitation could lead to an attacker gaining limited access to the application. Exploiting vulnerabilities may require
an attacker to manipulate users to gain access to their credentials. Such exploitation could lead to limited data loss
or downtime.

### Low

Exploitation will likely have very little impact on the application, and it is unlikely that an attacker will gain any
meaningful access to the application. Exploiting an issue of this severity will potentially require physical access to
the infrastructure that supports the application.

### Informational

While not part of the CVSS scoring specification, several security analysis tools use this severity level to indicate
that an issue is a matter of best practice. It is extremely unlikely that issues with this severity will lead to an
attacker gaining access to any application components.

### Unknown

This severity level is used when the analysis tool used to perform a scan of the application does not associate any kind
of severity level with the issues or vulnerabilities it finds. Issues with an unknown severity should be investigated by
application developers and project stakeholders to establish the ease of exploitation, scope of any potential impact and
the specific risks associated.
"
`;
